src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,557 @@
+
+
+def _process_input(conf: Conf) -> None:
+    """
+    Process input from input_queue and put the result in output_queue.
+    """
+    while True:
+        task = conf.input_queue.get(block=True)
+        if not task:
+            print("Stopping")
+            break
+
+        upa = None
+        (
+            upload_env_key,
+            workspace_name,
+            workspace_id,
+            assembly_path,
+            assembly_dir,
+            assembly_name,
+            upload_dir,
+            counter,
+            assembly_files_len,
+        ) = task
+
+        if counter % 3000 == 0:
+            print(f"{counter / assembly_files_len * 100:.2f}% assemblies have been processed")
+
+        try:
+            upa = _upload_assembly_to_workspace(conf, workspace_name, assembly_path, assembly_name)
+            _post_process(
+                upload_env_key,
+                workspace_id,
+                assembly_dir,
+                assembly_name,
+                upload_dir,
+                conf.output_dir,
+                upa
+            )
+        except Exception as e:
+            print(f"Failed assembly name: {assembly_name}. Exception:")
+            print(e)
+        conf.output_queue.put((assembly_name, upa))
+
+
+def _upload_assembly_files_in_parallel(
+        conf: Conf,
+        upload_env_key: str,
+        workspace_name: str,
+        workspace_id: int,
+        upload_dir: str,
+        temp_dir: str,
+        wait_to_upload_assemblies: dict[str, str],
+) -> list[str]:
+    """
+    Upload assembly files to the target workspace in parallel using multiprocessing.
+
+    Parameters:
+        conf: Conf object
+        upload_env_key: environment variable key in uploaded.yaml file
+        workspace_name: taregt workspace name
+        workspace_id: target workspace id
+        upload_dir: a directory in collectionssrouce that creates new directories linking to sourcedata
+        temp_dir: a temporary directory that stores the assembly files to upload
+        wait_to_upload_assemblies: a dictionary that maps assembly file name to assembly directory
+
+    Returns:
+        a list of assembly names that failed to upload
+    """
+    assembly_files_len = len(wait_to_upload_assemblies)
+    print(f"Start uploading {assembly_files_len} assembly files with {conf.workers} workers\n")
+
+    # Put the assembly files in the input_queue
+    counter = 1
+    for assembly_name, assembly_dir in wait_to_upload_assemblies.items():
+
+        if counter % 5000 == 0:
+            print(f"{counter / assembly_files_len * 100:.2f}% jobs have been added to the queue {datetime.now()}")
+
+        assembly_path = os.path.join(JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, temp_dir, assembly_name)
+        conf.input_queue.put(
+            (
+                upload_env_key,
+                workspace_name,
+                workspace_id,
+                assembly_path,
+                assembly_dir,
```

Comment:
```suggestion
                container_internal_assembly_path,
                host_assembly_dir,
```

Hard to understand what these vars are. Same changes in _process_input

Reply:
updated!

Reply:
:+1: